### PR TITLE
Fix #29390, add space after tuple type and identifier

### LINKED ISSUE #29390: NormalizeWhitespace do not add space after tuple type in parameter and in return
**Version Used**: 
The same behavior on *2.8.2*, *2.9.0*

**Steps to Reproduce**:

1. Create method declaration:
```csharp
static void Main(string[] args)
{
    var methodDeclaration = SyntaxFactory.MethodDeclaration(
        new SyntaxList<AttributeListSyntax>(),
        SyntaxFactory.TokenList(SyntaxFactory.Token(SyntaxKind.PublicKeyword)),
        SyntaxFactory.ParseTypeName("(string prefix, string uri)"),
        null,
        SyntaxFactory.Identifier("XmlPoke"),
        null,
        SyntaxFactory.ParameterList(
            SyntaxFactory.SeparatedList(new[]
            {
                SyntaxFactory.Parameter(
                    new SyntaxList<AttributeListSyntax>(),
                    SyntaxFactory.TokenList(),
                    SyntaxFactory.ParseTypeName("(string prefix, string uri)"),
                    SyntaxFactory.Identifier("ns"),
                    null)
            })),
        SyntaxFactory.List(new List<TypeParameterConstraintClauseSyntax>()),
        null,
        null
    );

    Console.WriteLine(methodDeclaration);
    Console.WriteLine(methodDeclaration.NormalizeWhitespace());
}
```

2. Run it

**Expected Behavior**:
Add whitespace between name and tuple type in return statement and in parameter definition:
```csharp
public (string prefix, string uri) XmlPoke((string prefix, string uri) ns)
```

**Actual Behavior**:
There are missed whitespaces (github even can't highlight this code correctly).
```csharp
public (string prefix, string uri)XmlPoke((string prefix, string uri)ns)
```

I don't think that anyone expect such behavior.
If we replace `tuple` with array of tuple that everything will be correct:

```csharp
public (string prefix, string uri)[] XmlPoke((string prefix, string uri)[] ns)
```

### PR DESCRIPTION
**Pull request by werwolfby**
Fix for #29390

But I don't know what to do with this:
```csharp
var a = new(string a, int b)[10];
```

From my point of view it should have space:
```csharp
var a = new (string a, int b)[10];
```

Completely similar to:
```csharp
var a = new string[10];
var b = new Lazy<int>[10];
// etc.
```

We don't have spaces only in 2 cases:
```csharp
class Foo<T>
    where T : new() // case #1
{
    public Foo()
    {
        var a = new[2] { 1, 2 }; // case #2
    }
}
```

In both cases we don't specify any type here, so skip space looks natural.
But for tuple from my point of view it is unnatural, because tuple - is just a type, and for types we have extract space.

I've found a few discussions:
from @jcouv: https://github.com/dotnet/csharplang/issues/100#issuecomment-350823105
and @sharwell : https://github.com/dotnet/roslyn/pull/23686#discussion_r156090382

But I can't find final decision

Right now I'll leave test code commented. And will not add whitespace before tuple in `new` expression. I need your input on this.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
> From my point of view it should have space:

I agree.  it should have a space.   When there's a type, it's `new<space>type`.

File: src/Compilers/CSharp/Portable/Syntax/SyntaxNormalizer.cs [hunk 1]
```diff
@@ -402,6 +402,12 @@ private static bool NeedsSeparator(SyntaxToken token, SyntaxToken next)
                 return true;
             }
 
+            // We don't want to add extra space after cast, we want space only after tuple
+            if (token.IsKind(SyntaxKind.CloseParenToken) && IsWord(next.Kind()) && token.Parent?.IsKind(SyntaxKind.TupleType) == true)
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Syntax/SyntaxNormalizer.cs]: 1. i don't understand what the IsWord check is for.  
2. you don't need `?.` after token.Parent.  All tokens you could reach in SyntaxNormalizer have parents.
3. I'd much rather we enumerate the cases where a tuple type would want a space after it (or where it woudl not want a space after it).  Looking specifically for `isword` just seems busted.

Comment (werwolfby) [reply src/Compilers/CSharp/Portable/Syntax/SyntaxNormalizer.cs]: In cases like:
```csharp
(string a, (string b, string c))
```
I don't want to add space after `c)` so I want to have space only after `)<identifier,word>` the same expression exist above but for braket: https://github.com/dotnet/roslyn/blob/ec99ff789c32a34fe444f8631e5a5166d2ebc11d/src/Compilers/CSharp/Portable/Syntax/SyntaxNormalizer.cs#L400

Comment (werwolfby) [reply src/Compilers/CSharp/Portable/Syntax/SyntaxNormalizer.cs]: So I think we want space after tuple only when next token is word.

**Comment by werwolfby** [comment]
Add space after new and before tuple:

```csharp
var a = new(string a, int b)[10];
```

Will be normalized to:

```csharp
var a = new (string a, int b)[10];
```

**Comment by werwolfby** [comment]
@jcouv Can we point it to v15.9 (or earlier) milestone?

**Comment by CyrusNajmabadi** [comment]
I'm not sure this would meet any 15.9 bar...

**Comment by werwolfby** [comment]
From my point of view, it is too small changes to postpone it.

But I'm ok with it.

**Comment by CyrusNajmabadi** [comment]
> P.S. By the way I can't understand your branching strategy: dev15.9.x,

'dev15.9.x' is the branch going into VS15.9.  It contains all the changes roslyn wants to go in that release, and should only be taking PRs that have been specifically approved for that release based on a high 'bar' check that warrants it.

'master' is where pretty much all other work goes.  At some point master will branch off onto whatever come after 'dev15.9.x'.  That may be 'dev15.10.x', 'dev16.x', or something else entirely.    At that point, that branch will again only be for approved changes for that specific release, and 'master' will again be for the release after that.

And so on and so forth.  Effectively: we do work in master.  And, from time to time, branches are made to capture the point in time containing all the work we want to go in a specific VS release.

**Comment by jcouv** [comment]
@ivanbasov for second review. Thanks

**Comment by jcouv** [comment]
Closed/re-opened to kick CI.

**Comment by jcouv** [comment]
Merged. Thanks @werwolfby !

**Pull request merged**